docs/walkthroughs/upgrading-v2-to-v3.md

```diff
@@ -0,0 +1,301 @@
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: kuard
+  labels:
+    app: kuard
+spec:
+  selector:
+    matchLabels:
+      app: kuard
+  template:
+    metadata:
+      labels:
+        app: kuard
+        marin3r.3scale.net/status: "enabled"
+      annotations:
+        marin3r.3scale.net/node-id: kuard
+        marin3r.3scale.net/ports: envoy-http:8081
+        marin3r.3scale.net/envoy-api-version: v2
+        marin3r.3scale.net/shutdown-manager.enabled: "true"
+        marin3r.3scale.net/envoy-image: envoyproxy/envoy:v1.16.4
+    spec:
+      containers:
+        - name: kuard
+          image: gcr.io/kuar-demo/kuard-amd64:blue
+          ports:
+            - containerPort: 8080
+              name: http
+              protocol: TCP
+EOF
+```
+
+You should see the following pods:
+
+```bash
+▶ kubectl get pods
+NAME                                        READY   STATUS    RESTARTS   AGE
+kuard-7dfdbb5656-vhtxk                      3/3     Running   0          61s
+marin3r-discoveryservice-7b4c6d4c7c-fftg2   1/1     Running   0          90m
+```
+
+### Update the EnvoyConfig to v3
+
+In order to upgrade to v3, you just need to set the `spec.envoyAPI` field to `v3` and update the syntax of your envoy resources to match the v3 API spec.
+
+```bash
+cat <<'EOF' | kubectl apply -f -
+apiVersion: marin3r.3scale.net/v1alpha1
+kind: EnvoyConfig
+metadata:
+  name: envoy
+spec:
+  nodeID: envoy
+  serialization: yaml
+  envoyAPI: v3
+  envoyResources:
+    listeners:
+      - name: http
+        value: |
+          name: http
+          address: { socket_address: { address: 0.0.0.0, port_value: 8080 } }
+          filter_chains:
+            - filters:
+                - name: envoy.filters.network.http_connection_manager
+                  typed_config:
+                    "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
+                    stat_prefix: ingress_http
+                    route_config:
+                      name: route
+                      virtual_hosts:
+                        - name: any
+                          domains: ["*"]
+                          routes:
+                            - {"match":{"prefix":"/"},"direct_response":{"status":200}}
+                    http_filters:
+                      - name: envoy.filters.http.router
+EOF
+```
+
+Check that the EnvoyConfig has been updated to v3:
+
+```bash
+▶ kubectl get envoyconfig
+NAME    NODE ID   ENVOY API   DESIRED VERSION   PUBLISHED VERSION   CACHE STATE
+envoy   envoy     v3          76795866f         76795866f           InSync
+```
+
+### Update the Envoy sidecars to use v3
+
+At this point, we have our EnvoyConfig updated to v3 but our Envoy sidecars still use v2. To do so, we update the `marin3r.3scale.net/envoy-api-version` annotation in the `kuard` Deployment. This will trigger a rollout of the Deployment and v3 sidecars will be injected into the new pods.
```

Comment:
At this point with EC with v3, the envoy image is still using 1.16:
- Maybe this envy image field should be removed from deployment (and so, starting to use the one that comes with the operator, on that case I guess 1.18
- Or at least, change the envoy image manually to use 1.18?

Reply:
As discussed in our daily stadup, upgrading the API and upgrading the Envoy version are different tasks that should be performed separately.